### ISSUE TITLE
Implementation of loss function depending on more than single output column

### ISSUE THREAD

**Issue by KOLANICH** [dmlc/xgboost#4556]
When doing multi-output regression loss often depends on another output column. So we need to get that column be excluded from features, but be available when computing loss.

For a single fitting it is not always a big problem - one just use the etalon from an initial array rather than from args of a function. But for crossvalidation it is not an option because XGBoost python lib splits and permutes DMatrices itself. sklearn crossvalidation is also not an option, it unneedingly tosses the matrices back and forth, it is damn slow.

The solution may be to get indices of permutted rows in custom loss func, but  ... it doesn't solve the problems of that issue preventing from implementing such losses in C++.

So we need a way to exclude certain columns from features and a way to get and use them in a loss function.

**Comment by trivialfis** [comment]
Could you give the specific name or link to paper of that algorithm?

**Comment by KOLANICH** [comment]
What algorithm do you mean? It is not from a paper, I just trying to make a multioutput regression by fitting a separate model for each output column (α, β). But L is such, so ∂L/∂β and ∂^2L/∂β^2 depend on α strongly, so these terms cannot be just removed.

**Comment by pford221** [comment]
How do you do multi-output regression with gradient-boosted decision trees?

**Comment by KOLANICH** [comment]
>How do you do multi-output regression with gradient-boosted decision trees?

Currently I don't because of some problems (lot of nans even if none of the data or loss first and second partial derivatives contain any nans). Since XGBoost has no multioutput regression, the natural thing to try is to try predicting each column with a separate model trained using a custom loss function `L(α, β, α', β')` (primed vars are from dataset, non-primed are predicted by XGBoost), replacing non-predictable var with the value from the dataset, so for α the loss is `L(α, β', α', β')`.

**Comment by KOLANICH** [comment]
>How do you do multi-output regression with gradient-boosted decision trees?

Currently I don't because of some problems (lot of nans even if none of the data or loss first and second partial derivatives contain any nans). Since XGBoost has no multioutput regression, the natural thing to try is to try predicting each column with a separate model trained using a custom loss function `L(α, β, α', β')` (primed vars are from dataset, non-primed are predicted by XGBoost), replacing non-predictable var with the value from the dataset, so for α the loss is `L(α, β', α', β')`.

**Comment by pford221** [comment]
Sorry, but you're going to have explain what the parameters $\alpha$ and $\beta$ mean? Are those hyper parameters, leaf weights??

**Comment by KOLANICH** [comment]
No. They are parameters of Weibull distribution. The loss function is [Kullback-Leibler divergence between 2 Weibull distributions](https://arxiv.org/pdf/1310.3713.pdf) (one should be predicted by a model) with some regularization terms meant to prevent the arguments from going negative.

**Comment by hcho3** [comment]
@KOLANICH As part of #4491, I'm looking to add two extra attributes `label_lower_bound` and `label_upper_bound`, to express left- and interval-censored data. So adding another attribute for the extra output column may be a solution for your use case.

**Comment by trivialfis** [comment]
Closing as duplicated to multioutput regression feature request.  It's under the radar now.
